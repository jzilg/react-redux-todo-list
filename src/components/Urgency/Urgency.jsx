import React from 'react'
import PropTypes from 'prop-types'
import { getDurationBetweenDates } from '../../utils/helper'

const Urgency = ({ today, lastEvent, schedule }) => {
    if (!lastEvent || !schedule) {
        return (
            <span>
                ?
            </span>
        )
    }

    function getUrgency(duration) {
        let days = schedule - duration
        if (days < 0) {
            days = -(days)
        }

        const dayText = days === 1 ? 'Day' : 'Days'
        const timeText = duration > schedule ? 'late' : 'left'

        return days === 0 ? 'Do it today' : `${days} ${dayText} ${timeText}`
    }

    function getUrgencyColor(duration) {
        let color = ''

        if (duration !== schedule) {
            color = duration > schedule ? 'font-red' : 'font-green'
        }

        return color
    }

    const duration = getDurationBetweenDates(today, lastEvent)
    const color = getUrgencyColor(duration, schedule)
    const urgency = getUrgency(duration, schedule)

    return (
        <span className={color}>
            {urgency}
        </span>
    )
}

Urgency.propTypes = ({
    today: PropTypes.string.isRequired,
    lastEvent: PropTypes.string.isRequired,
    schedule: PropTypes.number.isRequired,
})

export default Urgency
