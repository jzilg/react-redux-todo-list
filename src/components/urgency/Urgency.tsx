import React, { ReactElement } from 'react'
import { getDurationBetweenDates } from '../../utils/date'

interface Props {
    today: string
    lastEvent: string
    schedule: number
}

function Urgency(props: Props): ReactElement {
    const { today, lastEvent, schedule } = props

    if (!lastEvent || !schedule) {
        return (
            <span>
                ?
            </span>
        )
    }

    function getUrgency(duration: number): string {
        let days = schedule - duration
        if (days < 0) {
            days = -(days)
        }

        const dayText = days === 1 ? 'Day' : 'Days'
        const timeText = duration > schedule ? 'late' : 'left'

        return days === 0 ? 'Do it today' : `${days} ${dayText} ${timeText}`
    }

    function getUrgencyColor(duration: number): string {
        let color = ''

        if (duration !== schedule) {
            color = duration > schedule ? 'font-red' : 'font-green'
        }

        return color
    }

    const duration: number = getDurationBetweenDates(today, lastEvent)
    const color: string = getUrgencyColor(duration)
    const urgency: string = getUrgency(duration)

    return (
        <span className={color}>
            {urgency}
        </span>
    )
}

export default Urgency
