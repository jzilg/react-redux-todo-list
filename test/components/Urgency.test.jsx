import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { shallowToJson } from 'enzyme-to-json'
import expect from 'expect'
import Urgency from '../../src/components/Urgency'

Enzyme.configure({ adapter: new Adapter() })

describe('Urgency', () => {
    const pastOneDayAgo = '2018-09-16'
    const pastMoreDaysAgo = '2018-09-14'
    const today = '2018-09-17'
    const schedule = 1

    it('should render "do it today" if lastEvent was one day ago and schedule is 1', () => {
        const output = shallow((
            <Urgency
                today={today}
                lastEvent={pastOneDayAgo}
                schedule={schedule}
            />
        ))
        expect(shallowToJson(output)).toMatchSnapshot()
    })

    it('should render "n Days left" if was n days ago', () => {
        const output = shallow((
            <Urgency
                today={today}
                lastEvent={pastMoreDaysAgo}
                schedule={schedule}
            />
        ))
        expect(shallowToJson(output)).toMatchSnapshot()
    })

    it('should render "1 day left" if lastEvent was today', () => {
        const output = shallow((
            <Urgency
                today={today}
                lastEvent={today}
                schedule={schedule}
            />
        ))
        expect(shallowToJson(output)).toMatchSnapshot()
    })

    it('should render "?" if last event is not set', () => {
        const output = shallow((
            <Urgency
                today={today}
                lastEvent=""
                schedule={schedule}
            />
        ))
        expect(shallowToJson(output)).toMatchSnapshot()
    })
})
