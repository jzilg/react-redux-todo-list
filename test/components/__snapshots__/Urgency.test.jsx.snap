// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Urgency should render "?" if last event is not set 1`] = `
<span>
  ?
</span>
`;

exports[`Urgency should render "1 day left" if lastEvent was today 1`] = `
<span
  className="font-green"
>
  1 Day left
</span>
`;

exports[`Urgency should render "do it today" if lastEvent was one day ago and schedule is 1 1`] = `
<span
  className=""
>
  Do it today
</span>
`;

exports[`Urgency should render "n Days left" if was n days ago 1`] = `
<span
  className="font-red"
>
  2 Days late
</span>
`;
